cli/tests/test_squash_command.rs

```diff
@@ -749,6 +749,742 @@ fn test_squash_from_to_partial() {
     ");
 }
 
+#[test]
+fn test_squash_working_copy_restore_descendants() {
+    let test_env = TestEnvironment::default();
+    test_env.run_jj_in(".", ["git", "init", "repo"]).success();
+    let work_dir = test_env.work_dir("repo");
+
+    // Create history like this:
+    //
+    // B C@
+    // |/
+    // A
+    //
+    // Each commit adds a file named the same as the commit
+    let create_commit = |name: &str| {
+        work_dir
+            .run_jj(["bookmark", "create", "-r@", name])
+            .success();
+        work_dir.write_file(name, format!("test {name}\n"));
+    };
+
+    create_commit("a");
+    work_dir.run_jj(["new"]).success();
+    create_commit("b");
+    work_dir.run_jj(["new", "a"]).success();
+    create_commit("c");
+
+    let template = r#"separate(
+        " ",
+        commit_id.short(),
+        bookmarks,
+        description,
+        if(empty, "(empty)")
+    )"#;
+    let run_log = || work_dir.run_jj(["log", "-r=::", "--summary", "-T", template]);
+
+    // Verify the setup
+    insta::assert_snapshot!(run_log(), @r"
+    @  301231f4040a c
+    │  A c
+    │ ○  a67b55091cc7 b
+    ├─╯  A b
+    ○  b73559961e51 a
+    │  A a
+    ◆  000000000000 (empty)
+    [EOF]
+    ");
+    let output = work_dir.run_jj(["file", "list", "-r=a"]);
+    insta::assert_snapshot!(output, @r"
+    a
+    [EOF]
+    ");
+    let output = work_dir.run_jj(["file", "list", "-r=b"]);
+    insta::assert_snapshot!(output, @r"
+    a
+    b
+    [EOF]
+    ");
+    let output = work_dir.run_jj(["file", "list"]);
+    insta::assert_snapshot!(output, @r"
+    a
+    c
+    [EOF]
+    ");
+
+    let output = work_dir.run_jj(["squash", "--restore-descendants"]);
+    insta::assert_snapshot!(output, @r"
+    ------- stderr -------
+    Working copy now at: kpqxywon 3cb3a3a7 (no description set)
+    Parent commit      : qpvuntsm abd9e679 a c | (no description set)
+    Added 0 files, modified 0 files, removed 1 files
+    [EOF]
+    ");
+
+    let output = work_dir.run_jj(["show"]);
+    //  BUG: The current commit should become empty!
```

Comment:
I found a bug: if you do jj new X; <edit X>; jj squash --restore-descendants, the commit created for the working copy is not empty.

It's not obvious to me how to solve it, would appreciate advice.

Reply:
My current plan is to combine `test_reparent_descendants` in test_mut_repo with the working copy tests in the same file, since I suspect the bug is in `reparent_descendants`.

Reply:
Perhaps, this is because we do `reparent` in `squash_commits()` before the destination commit is rewritten. The new wc commit would be created on top of the original `destination` commit, then reparented.

Reply:
I couldn't reproduce the other bug I was worrying about in the previous message. Not sure how, but everything seems to be working correctly now  AFAICT :)